client/verta/verta/client.py

```diff
@@ -3360,7 +3360,81 @@ def get_artifact(self, key):
             except:
                 return six.BytesIO(artifact)
 
-    # TODO: download_artifact(self, key, download_to_path)
+    def download_artifact(self, key, download_to_path, chunk_size=32*(10**3)):
+        """
+        Downloads the artifact with name `key` to path `download_to_path`.
+
+        Parameters
+        ----------
+        key : str
+            Name of the artifact.
+        download_to_path : str
+            Path to download to.
+        chunk_size : int, default 32 kB
+            Number of bytes to download at a time.
+
+        Returns
+        -------
+        downloaded_to_path : str
+            Absolute path where artifact was downloaded to. Matches `download_to_path`.
+
+        """
+        download_to_path = os.path.abspath(download_to_path)
+
+        # get key-path from ModelDB
+        # TODO: consolidate the following ~12 lines with ExperimentRun._get_artifact()
+        Message = _CommonService.GetArtifacts
+        msg = Message(id=self.id, key=key)
+        data = _utils.proto_to_json(msg)
+        response = _utils.make_request("GET",
+                                       "{}://{}/api/v1/modeldb/experiment-run/getArtifacts".format(self._conn.scheme, self._conn.socket),
+                                       self._conn, params=data)
+        _utils.raise_for_http_error(response)
+
+        response_msg = _utils.json_to_proto(_utils.body_to_json(response), Message.Response)
+        artifact = {artifact.key: artifact for artifact in response_msg.artifacts}.get(key)
+        if artifact is None:
+            raise KeyError("no artifact found with key {}".format(key))
+
+        # TODO: unpack dirs logged as artifacts
+        #     But we can't distinguish if a ZIP artifact is a directory we've compressed, or if it
+        #     was a ZIP file the user already had.
+
+        # create parent dirs
+        pathlib2.Path(download_to_path).parent.mkdir(parents=True, exist_ok=True)
+        # TODO: clean up empty parent dirs if something later fails
+
+        # get a stream of the file bytes, without loading into memory, and write to file
+        # TODO: consolidate this with _get_artifact() and get_artifact()
+        print("downloading {} from ModelDB".format(key))
+        if artifact.path_only:
+            if os.path.exists(artifact.path):
+                # copy from clientside storage
+                # TODO: use a tempfile first, and also delete if failed
+                with open(artifact.path, 'rb') as src_f:
+                    with open(download_to_path, 'wb') as dest_f:
+                        shutil.copyfileobj(src_f, dest_f)
```

Comment:
Why not `copyfile`?

Reply:
I think I just really keep gravitating towards using streams.

Also, my understanding was that `copyfile` just uses `copyfileobj` under the hood anyway, rather than an OS-level `cp`.

Actually, the Py3 docs say `copyfile` uses "the most efficient way possible", so I'll use that.

Reply:
If it doesn't use OS-level `cp`, I'd be surprised to say the least.